src/sinks/datadog/traces/apm_stats/flusher.rs

```diff
@@ -0,0 +1,173 @@
+            let mut aggregator = self.aggregator.lock().unwrap();
+            let client_stats_payloads = aggregator.flush(force);
+
+            if client_stats_payloads.is_empty() {
+                // no sense proceeding if no payloads to flush
+                None
+            } else {
+                let payload = StatsPayload {
+                    agent_hostname: aggregator.get_agent_hostname(),
+                    agent_env: aggregator.get_agent_env(),
+                    stats: client_stats_payloads,
+                    agent_version: aggregator.get_agent_version(),
+                    client_computed: false,
+                };
+
+                Some((payload, aggregator.get_api_key()))
+            }
+        } {
+            if let Err(e) = self.compress_and_send(payload, api_key).await {
+                // TODO emit an internal `Error` event here, probably
+                error!(
+                    message = format!("Error while encoding and sending APM stats payloads: {}", e),
+                    internal_log_rate_limit = true,
+                );
+            }
+        }
+    }
+
+    async fn compress_and_send(
+        &self,
+        payload: StatsPayload,
+        api_key: Arc<str>,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        let (metadata, compressed_payload) = self.build_apm_stats_request_data(api_key, payload)?;
+
+        let request_metadata = RequestMetadata::new(0, 0, 0, 0, 0);
+        let trace_api_request = build_request(
+            (metadata, request_metadata),
+            compressed_payload,
+            self.compression,
+            &self.endpoint_configuration,
+        );
+
+        let http_request = trace_api_request
+            .into_http_request()
+            .context(BuildRequestSnafu)?;
+
+        self.client.send(http_request).await?;
+
+        Ok(())
+    }
+
+    fn build_apm_stats_request_data(
+        &self,
+        api_key: Arc<str>,
+        payload: StatsPayload,
+    ) -> Result<(DDTracesMetadata, Bytes), RequestBuilderError> {
+        let encoded_payload =
+            rmp_serde::to_vec_named(&payload).map_err(|e| RequestBuilderError::FailedToEncode {
+                message: "APM stats encoding failed.",
+                reason: e.to_string(),
+                dropped_events: 0,
+            })?;
+        let uncompressed_size = encoded_payload.len();
+        let metadata = DDTracesMetadata {
+            api_key,
+            endpoint: DatadogTracesEndpoint::APMStats,
+            finalizers: EventFinalizers::default(),
+            uncompressed_size,
+            content_type: "application/msgpack".to_string(),
+        };
+
+        let mut compressor = Compressor::from(self.compression);
+        match compressor.write_all(&encoded_payload) {
+            Ok(()) => {
+                let bytes = compressor.into_inner().freeze();
+
+                Ok((metadata, bytes))
+            }
+            Err(e) => Err(RequestBuilderError::FailedToEncode {
```

Comment:
Is there/should there be a separate error for encoding vs compressing?

Reply:
Mmm. Yeah these bits were existing and moved. The Snafu for `RequestBuilderError` could easily distinguish them better.

And it would make more sense because in encoding failing case the Error message would be a bit redundant:

"Encoding of a request payload failed (APM stats encoding failed, (...error...))"

I will change it.

Reply:
So, I made changes to this area but instead of adding a separate Snafu to that enum, I just updated the strings for the messages so that it more clearly distinguished between compression and encoding.

Went down the other path but this Snafu error is used in the request_builder.rs itself and this seemed a less intrusive approach that still satisfies the goal. LMK if doesn't though.